great_expectations/experimental/datasources/file_path_data_asset.py

```diff
@@ -16,21 +20,31 @@
     DataAsset,
 )
 
+if TYPE_CHECKING:
+    from great_expectations.experimental.datasources import (
+        PandasFilesystemDatasource,
+        SparkDatasource,
+    )
+
 logger = logging.getLogger(__name__)
 
 
 class _FilePathDataAsset(DataAsset):
     _EXCLUDE_FROM_READER_OPTIONS: ClassVar[Set[str]] = {
         "name",
-        "regex",
-        "order_by",
+        "regex_setter",
+        "order_by_setter",
         "type",
     }
 
     # General file-path DataAsset pertaining attributes.
-    regex: Pattern
+    regex_setter: Union[str, Pattern] = pydantic.Field(alias="regex")
```

Comment:
If we don't use an alias here, `mypy` will complain that you can't set the `regex` property. This enables us to pass along the typing information from here to the API-level.

Reply:
The order of types here needs to be swapped.
Pydantic will attempt to parse types according to a union in the order the union is defined, so almost any object will be coercible into a string and then pydantic will just stop there and not attempt to parse it into the more specific type.
https://docs.pydantic.dev/usage/types/#unions

Why do we want to make this a `str | Pattern` anyways? Is this just because mypy complains about regex strings?

Reply:
Yes, this is just about `mypy`. We could have different types for `add_csv_asset()` `regex` and `CSVAsset.regex` like it was before, but then we should do type conversion (instead of `type: ignore`) in `add_csv_asset` instead of in the `regex` property below here. I opted to handle it internally, abstracting this away from the end user/contributor.

[According to pydantic](https://docs.pydantic.dev/usage/types/) they just use `re.compile` for coercion, so it's not like we are getting something special by having them do the type conversion.

```
typing.Pattern
will cause the input value to be passed to re.compile(v) to create a regex pattern
```

Reply:
I'm happy to do a `re.compile` in each `add_*_asset` method if we would prefer that. I just don't like the `type: ignore`.

Reply:
>I'm happy to do a re.compile in each add_*_asset method if we would prefer that. I just don't like the type: ignore.

Yes, let's do that instead of having property setters/getters.

Reply:
✅